### PR TITLE
Implementation of cosine learning rate training policy

### PR DESCRIPTION
**Pull request by grantlj**
Summary:
Implementation of the cosine learning rate in: https://arxiv.org/pdf/1608.03983.pdf.

Mostly inspired from:
https://github.com/pytorch/fairseq/blob/master/fairseq/optim/lr_scheduler/cosine_lr_scheduler.py

Test Plan:
buck test -v 2 caffe2/caffe2/fb/dper/layer_models/tests/split_1:sparse_nn_test  -- test_composite_cosine_lr_policy

learning rate log with max_lr=0.3, initial_period=20, t_mult=0.95, lr_shrink=0.95: P120327179

https://pxl.cl/PrcP

full canary: https://fburl.com/fblearner/mw69ylsd

Differential Revision: D18195868

### CODE REVIEW

**Comment by peterjc123** [comment]
Looks like  `_USE_MATH_DEFINES` isn't actually working on Windows. Could you please define `M_PI` manually when it is not available?

**Comment by zou3519** [comment]
This broke the windows build so we reverted the change. See @peterjc123's comment.
